### PR TITLE
Add code coverage

### PR DESCRIPTION
**Pull request by torkleyy**
## Description

Make Jenkins collect code coverage reports and upload them to https://codecov.io/

Credit goes to @fhaynes, just opening the PR.

## PR Checklist

By placing an x in the boxes I certify that I have:

- [ ] Ran `cargo test --all` locally if this modified any rs files.
- [ ] Ran `cargo +stable fmt --all` locally if this modified any rs files.
- [ ] Updated the content of the book if this PR would make the book outdated.
- [ ] Added a changelog entry if this will impact users, or modified more than 5 lines of Rust that wasn't a doc comment.
- [ ] Added unit tests for new APIs if any were added in this PR.
- [x] Acknowledged that by making this pull request I release this code under an MIT/Apache 2.0 dual licensing scheme.

### CODE REVIEW

File: Jenkinsfile [hunk 1]
```diff
@@ -88,6 +88,31 @@ pipeline {
                         echo 'Tests done!'
                     }
                 }
+                stage('Calculate Coverage') {
+                    environment {
+                        CARGO_HOME = '/home/jenkins/.cargo'
```
Comment (LucioFranco) [inline Jenkinsfile]: you don't need these two iirc on unix.

File: Jenkinsfile [hunk 2]
```diff
@@ -88,6 +88,26 @@ pipeline {
                         echo 'Tests done!'
                     }
                 }
+                stage('Calculate Coverage') {
```
Comment (LucioFranco) [inline Jenkinsfile]: we should see if we can move this to its own stage? so its

check/rustfmt then tests then coverage

Comment (torkleyy) [reply Jenkinsfile]: We can, but what's the advantage of that? Wouldn't that just increase the build time?

Comment (LucioFranco) [reply Jenkinsfile]: only reason I see is if codecov ratelimits us and we prob dont wanna post coverage unless tests pass?

Comment (LucioFranco) [reply Jenkinsfile]: but I dont mind either way.

Comment (torkleyy) [reply Jenkinsfile]: I don't think it's gonna be a problem, so let me revert the broken commit I just pushed.

Comment (LucioFranco) [inline Jenkinsfile]: Minor nit: maybe just `Coverage` here is better? 🤷‍♂

**Pull request merged**